inginious/agent/docker_agent/__init__.py

```diff
@@ -77,20 +116,19 @@ async def _init_clean(self):
         # Docker
         self._docker = AsyncProxy(DockerInterface())
 
+        if self._runtimes is None:
+            self._runtimes = self._detect_runtimes()
+
         # Auto discover containers
         self._logger.info("Discovering containers")
-        self._containers = await self._docker.get_containers()
-        for idx in self._containers:
-            self._containers[idx]["type"] = "docker" # type is not given by self._docker.get_containers()
-
-        self._assigned_external_ports = {}  # container_id : [external_ports]
+        self._containers = await self._docker.get_containers(self._runtimes.values())
 
         if self._address_host is None and len(self._containers) != 0:
             self._logger.info("Guessing external host IP")
-            self._address_host = await self._docker.get_host_ip(next(iter(self._containers.values()))["id"])
+            self._address_host = await self._docker.get_host_ip(next(iter(next(iter(self._containers.values())).values()))["id"])
```

Comment:
It would be useful to add some comment and explain why such a lot of next/iter :sweat_smile: